### PR TITLE
feat: Full HIP-1137 implementation

### LINKED ISSUE #23462: HIP-1137: Block Node Discoverabilty
### Background

https://hips.hedera.com/hip/hip-1137

### Acceptance Criteria

Sub-issues implemented

### Dependencies

_No response_

### Definition of Ready (DoR) Checklist

- [ ] Clear acceptance criteria
- [ ] Clear and detailed description
- [ ] Dependencies identified
- [ ] Links to documentation
- [ ] Should be completable in 2-3 Days
- [ ] Initial draft of Low-level design document
- [ ] At least high level test plan
- [ ] Groomed/Estimated

### Definition of Done (DoD) Checklist

- [ ] Acceptance Criteria complete
- [ ] No Codacy issues greater than minor (in new code)
- [ ] JavaDocs updated/created
- [ ] Code commented
- [ ] Unit tests created/updated
- [ ] 80% test code coverage (in new code)
- [ ] Happy Path and major negative cases in HAPI tests as applicable

### PR DESCRIPTION
**Pull request by petreze**
**Description**:
[HIP-1137](https://hips.hedera.com/hip/hip-1137)

**New Transaction Types**
- `RegisteredNodeCreate` — Creates a registered node with admin key, description, and typed service endpoints (block node / mirror node / RPC relay).
- `RegisteredNodeUpdate` — Selectively updates admin key, description, or service endpoints of an existing registered node. Requires existing admin key signature; if rotating the key, requires both old and new key signatures.
- `RegisteredNodeDelete` — Deletes a registered node. Protected by association check: fails with `REGISTERED_NODE_STILL_ASSOCIATED` if any non-deleted consensus node references it in its associatedRegisteredNode list. Treasury, system admin, and address book admin can delete without the node’s admin key.

**Consensus Node Integration**
- `NodeCreate` and `NodeUpdate` handlers now accept an `associatedRegisteredNode` list, allowing consensus nodes to declare which registered nodes they are associated with.
- Association list is validated: each referenced registered node must exist, and the list size must not exceed `maxAssociatedRegisteredNodes`.

**State & Schema**
- New `V073AddressBookSchema` adds a `REGISTERED_NODES` virtual map state (EntityNumber → RegisteredNode)
- `EntityCounts` proto extended with `num_registered_nodes` field.

**Protobuf Changes**
- 6 new response codes (529–534): `INVALID_REGISTERED_ENDPOINT`, `REGISTERED_ENDPOINTS_EXCEEDED_LIMIT`, `INVALID_REGISTERED_ENDPOINT_ADDRESS`, `INVALID_REGISTERED_ENDPOINT_TYPE`, `REGISTERED_NODE_STILL_ASSOCIATED`, `MAX_REGISTERED_NODES_EXCEEDED`.
-New state identifiers in `virtual_map_state.proto` for registered nodes (56).
-New singleton update field node_id_value in `state_changes.proto` for block stream tracking.

**Validation**
- Service endpoint validation: port range (0–65535), IP address format (4-byte IPv4 or 16-byte IPv6), FQDN validation (ASCII, label length, DNS rules).
- Endpoint type enforcement: exactly one of block_node, mirror_node, or rpc_relay must be set.
- Description validation: max UTF-8 byte length, no null bytes.

**App Wiring**
- `TransactionDispatcher`, `TransactionHandlers`, `ServiceScopeLookup`, `HandleWorkflowModule` — routed and wired for all three new transaction types.
- `ReadableStoreFactoryImpl` / `WritableStoreFactory` — registered `ReadableRegisteredNodeStore` and `WritableRegisteredNodeStore`.

**Block Stream**
- `BlockStreamBuilder` and `RecordStreamBuilder` implement `RegisteredNodeCreateStreamBuilder` to track registered node IDs in receipts.
- `BlockItemsTranslator` handles `REGISTERED_NODE_CREATE` via `NodeOpContext`.

**Fee, Throttle & Permissions**
- Fee schedules updated across all networks (mainnet, testnet, previewnet, genesis)
- Throttle definitions updated: all three operations added to network admin throttle group; `RegisteredNodeCreate` added to account creation throttle (2000 milliOps/sec).
- Scheduling whitelist updated to include all three transaction types.
- API permission properties updated across all environments.

**Testing**
- Unit tests: Full coverage for all three handlers, both stores, V073 schema, validators (including FQDN edge cases), fee calculators, and service definition.
- App-layer tests: Updated `BlockItemsTranslatorTest`, `DispatchHandleContextTest`, `WritableStoreFactoryImplTest`, `PreHandleWorkflowImplTest`, `PlatformStateUpdatesTest`.
- End-to-end `HapiSpec` tests: `RegisteredNodeTest` with comprehensive CRUD scenarios, error case validation, atomic batch tests, governance transaction tests and scheduling tests.
- Test infrastructure: New `HapiRegisteredNodeCreate/Update/Delete` DSL operations, `RegisteredNodeCreateTranslator`, registry support for registered node IDs, `TxnVerbs` string-based overloads.
- Bug fix: `FeeScheduleUpdateWaiverTest` now dynamically handles fee schedule chunk count instead of hard-coding 5 chunks (which broke when fee schedule grew past 20480 bytes with RegisteredNode entries).

- Added yahcli support for all of the new functionalities and tests to verify the behavior

**Related issue(s)**:

Fixes #23462 

**Notes for reviewer**:

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/WritableRegisteredNodeStore.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    protected WritableKVState<EntityNumber, RegisteredNode> registeredNodesState() {
+        return super.registeredNodesState();
+    }
+
+    public void put(@NonNull final RegisteredNode node) {
+        requireNonNull(node);
+        registeredNodesState()
+                .put(EntityNumber.newBuilder().number(node.registeredNodeId()).build(), node);
+    }
+
```
Comment (derektriley) [inline hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/WritableRegisteredNodeStore.java:36]: I believe we want a putAndIncrementCount method right to accurately track entity counts

File: hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/WritableRegisteredNodeStore.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    public void remove(final long registeredNodeId) {
+        registeredNodesState()
+                .remove(EntityNumber.newBuilder().number(registeredNodeId).build());
```
Comment (derektriley) [inline hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/WritableRegisteredNodeStore.java:57]: Would also need to decrement entityCounters here

**Comment by jsync-swirlds** [comment]
This is both out of date and cross-merged.
Perhaps consider running `git fetch && git rebase origin/main` instead of `git merge` to reduce the incidence of these broken merge events?
Alternatively, if you prefer git merge, try `git fetch origin && git merge origin/main` to at least bypass the outdated local main branch.

File: hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/validators/AddressBookValidator.java [hunk 3]
```diff
@@ -225,4 +259,112 @@ public static void validateX509Certificate(@NonNull final Bytes x509CertBytes) t
...
+            if (desc != null && !desc.isEmpty()) {
+                final var raw = desc.getBytes(StandardCharsets.UTF_8);
+                validateFalse(
+                        raw.length > nodesConfig.maxGeneralServiceDescriptionUtf8Bytes(), INVALID_REGISTERED_ENDPOINT);
+                validateFalse(containsZeroByte(raw), INVALID_REGISTERED_ENDPOINT);
+            }
+        }
+    }
+
+    private boolean isValidAsciiFqdn(@Nullable final String domain, final int maxFqdnSize) {
```
Comment (derektriley) [inline hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/validators/AddressBookValidator.java:322]: Registered endpoint FQDN validator accepts leading/trailing whitespace

isValidAsciiFqdn() trims input before validation. That means values like " block.example.com " are accepted, even though proto docs require a valid FQDN (whitespace should be invalid).
Because the original string is stored unchanged, this can persist malformed endpoint values

Comment (derektriley) [reply hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/validators/AddressBookValidator.java:322]: We should address this and add a test case

Comment (petreze) [reply hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/validators/AddressBookValidator.java:322]: Good catch! Added, plus some improvements to the fqdn validation 👍

File: hedera-node/data/config/application.properties [hunk 4]
```diff
@@ -2,7 +2,7 @@
 bootstrap.throttleDefsJson.resource=throttles-dev.json
 ledger.id=0x03
 staking.periodMins=1
-scheduling.whitelist=ConsensusSubmitMessage,CryptoTransfer,TokenMint,TokenBurn,CryptoApproveAllowance,CryptoUpdate
+scheduling.whitelist=ConsensusSubmitMessage,CryptoTransfer,TokenMint,TokenBurn,CryptoApproveAllowance,CryptoUpdate,RegisteredNodeCreate,RegisteredNodeUpdate,RegisteredNodeDelete
```
Comment (Neeharika-Sompalli) [inline hedera-node/data/config/application.properties:5]: Why are we adding these to schedules? Is this a requirement?

Comment (petreze) [reply hedera-node/data/config/application.properties:5]: Yes, the HIP states that these have to be schedulable transactions as well and added to the `SchedulableTransactionBody`

File: hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/calculator/RegisteredNodeCreateFeeCalculator.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+public class RegisteredNodeCreateFeeCalculator implements ServiceFeeCalculator {
+    @Override
+    public void accumulateServiceFee(
+            @NonNull final TransactionBody txnBody,
+            @NonNull final SimpleFeeContext simpleFeeContext,
+            @NonNull final FeeResult feeResult,
+            @NonNull final FeeSchedule feeSchedule) {
+        if (simpleFeeContext.feeContext() != null) {
+            checkDABEnabled(simpleFeeContext.feeContext());
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/calculator/RegisteredNodeCreateFeeCalculator.java]: Why are we checking DAB enabled here ? Isn't this a completely different transaction type. Are we saying the same property applies for these new transactions?

Comment (Neeharika-Sompalli) [reply hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/calculator/RegisteredNodeCreateFeeCalculator.java]: May be that makes sense.

File: hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/handlers/NodeUpdateHandler.java [hunk 6]
```diff
@@ -216,6 +228,10 @@ private Node.Builder updateNode(
         if (op.hasGrpcProxyEndpoint()) {
             nodeBuilder.grpcProxyEndpoint(unsetWebProxy ? null : op.grpcProxyEndpoint());
         }
+        if (op.hasAssociatedRegisteredNodeList()) {
+            nodeBuilder.associatedRegisteredNode(
+                    op.associatedRegisteredNodeListOrThrow().associatedRegisteredNode());
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/handlers/NodeUpdateHandler.java:233]: This name should be `associatedRegisteredNodes`  instead of `associatedRegisteredNode`

File: hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/handlers/RegisteredNodeDeleteHandler.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        final var registeredNodeId = op.registeredNodeId();
+        final var existingNode = registeredNodeStore.get(registeredNodeId);
+        validateFalse(existingNode == null, INVALID_REGISTERED_NODE_ID);
+
+        // Forbid deletion while associated by any consensus node.
+        final var isAssociated = nodeStore.keys().stream()
+                .map(key -> nodeStore.get(key.number()))
+                .anyMatch(node -> node != null
+                        && !node.deleted()
+                        && node.associatedRegisteredNode().contains(registeredNodeId));
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/handlers/RegisteredNodeDeleteHandler.java:87]: something to consider for future, any deletion we have to lookup all nodes. May be its useful to maintain in memory mappings. Not a blocker for this PR

File: hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/handlers/NodeCreateHandler.java
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/handlers/NodeCreateHandler.java:146]: Now these system dispatches for transplants should also look if the node id is not in registered node id list right?

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/junit/support/translators/BaseTranslator.java [hunk 8]
```diff
@@ -1021,6 +1022,11 @@ private void scanUnit(@NonNull final BlockTransactionalUnit unit) {
                         nextCreatedNums
                                 .computeIfAbsent(NODE, ignore -> new LinkedList<>())
                                 .add(nodeId);
+                    } else if (value.hasRegisteredNodeValue()) {
```
Comment (derektriley) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/junit/support/translators/BaseTranslator.java:1037]: registeredNodeId can be emitted in block-stream translation for mixed atomic batches
BaseTranslator records every registeredNode map update as a “created” ID, including updates to existing IDs; then RegisteredNodeCreateTranslator consumes that queue as if it only contains newly created IDs. In a batch containing both update(s) and create(s), this can attach the wrong ID to a REGISTERED_NODE_CREATE receipt.

I don't see any test that covers atomic batch with both registeredNodeUpdate and registeredNodeCreate and asserts receipt ID correctness in block translation, we should probably add this to verify.

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/HapiPropertySource.java [hunk 9]
```diff
@@ -455,6 +456,143 @@ static ServiceEndpoint asDnsServiceEndpoint(String v) {
...
+        boolean requiresTls = false;
+        for (int i = 2; i < parts.length; i++) {
+            final String part = parts[i];
+            if (part.equalsIgnoreCase("tls")) {
+                requiresTls = true;
+            } else {
+                try {
+                    blockNodeApi = RegisteredServiceEndpoint.BlockNodeEndpoint.BlockNodeApi.valueOf(part.toUpperCase());
+                } catch (IllegalArgumentException ignore) {
+                    // not a known api name; skip
```
Comment (derektriley) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/HapiPropertySource.java]: unknown trailing tokens are ignored instead of rejected. This allows typos to pass unnoticed (e.g., unexpected suffixes), which can lead to unintended endpoint config. We should probably fail the yahcli command

Comment (derektriley) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/HapiPropertySource.java]: There are no negative tests for yahcli endpoint parsing with unknown trailing segments (typos), e.g. extra non-tls tokens.

Comment (derektriley) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/HapiPropertySource.java]: No negative tests for malformed endpoint shapes with insufficient segments (addr without :port) in registered-node yahcli command paths.

File: hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/handlers/RegisteredNodeDeleteHandler.java (hunk 7 above)
Comment (petreze) [reply hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/handlers/RegisteredNodeDeleteHandler.java:87]: Agreed, good point for future optimization 👍

File: hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/handlers/NodeUpdateHandler.java (hunk 6 above)
Comment (petreze) [reply hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/handlers/NodeUpdateHandler.java:233]: The specified property in the HIP in the protos is `repeated uint64 associated_registered_node`, but maybe it makes more sense for plural

File: hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/calculator/RegisteredNodeCreateFeeCalculator.java (hunk 5 above)
Comment (petreze) [reply hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/calculator/RegisteredNodeCreateFeeCalculator.java]: It will be better to introduce a new one for HIP-1137 and not reuse the dab one. Will push an update

File: hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/handlers/NodeCreateHandler.java
Comment (petreze) [reply hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/handlers/NodeCreateHandler.java:146]: Correct, that's a nice catch, ty Neeha! Updated

File: hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/handlers/NodeUpdateHandler.java (hunk 6 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/handlers/NodeUpdateHandler.java:233]: yeah `associated_registered_node` returning a `List`. So we can change it to plural in next PR

Comment (petreze) [reply hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/handlers/NodeUpdateHandler.java:233]: I also see that we have followed this approach for other plural fields as well (e.g [1](https://github.com/hiero-ledger/hiero-consensus-node/blob/main/hapi/hedera-protobuf-java-api/src/main/proto/services/state/addressbook/node.proto#L67) and [2](https://github.com/hiero-ledger/hiero-consensus-node/blob/main/hapi/hedera-protobuf-java-api/src/main/proto/services/state/addressbook/node.proto#L80))

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/junit/support/translators/BaseTranslator.java (hunk 8 above)
Comment (petreze) [reply hedera-node/test-clients/src/main/java/com/hedera/services/bdd/junit/support/translators/BaseTranslator.java:1037]: Great catch, I will update BaseTranslator to cover those cases

File: hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/handlers/RegisteredNodeDeleteHandler.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+        final var registeredNodeId = op.registeredNodeId();
+        final var existingNode = registeredNodeStore.get(registeredNodeId);
+        validateFalse(existingNode == null, INVALID_REGISTERED_NODE_ID);
+
+        // Forbid deletion while associated by any consensus node.
+        final var isAssociated = nodeStore.keys().stream()
+                .map(key -> nodeStore.get(key.number()))
+                .anyMatch(node -> node != null
+                        && !node.deleted()
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-addressbook-service-impl/src/main/java/com/hedera/node/app/service/addressbook/impl/handlers/RegisteredNodeDeleteHandler.java:86]: I guess this is not possible to delete a node with a node associated right? So, I don't think we need `!node.deleted()`. Even though, it doesn't matter, it might create doubts in the future.

File: hapi/hedera-protobuf-java-api/src/main/proto/services/registered_service_endpoint.proto
Comment (Neeharika-Sompalli) [inline hapi/hedera-protobuf-java-api/src/main/proto/services/registered_service_endpoint.proto:39]: We also allow  non-FQDN hostnames like localhost right ?

Comment (petreze) [reply hapi/hedera-protobuf-java-api/src/main/proto/services/registered_service_endpoint.proto:39]: Correct

Comment (Neeharika-Sompalli) [reply hapi/hedera-protobuf-java-api/src/main/proto/services/registered_service_endpoint.proto:39]: I think we should add that in documentation in later PR
